### PR TITLE
Batch Fixes

### PR DESCRIPTION
**Pull request by mikepizzo**
1) Fix batch dispose issue (error writing to a closed stream)
2) Fix async batch issues (synchronous write to an asynchronous reader error)

### Issues

*This pull request fixes issue #2009*

### Description

In certain batch cases, the stream used to write the response was disposed before all the responses were written, resulting in a closed stream exception.

Also, certain async batch operations called synchronous operations under the covers, resulting in an exception regarding calling synchronous operations after an asynchronous operation.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    // Normally, the stream is closed when the ODataMessageWriter is disposed.
+    // For responses within a batch, we need to read from the stream after the response is
+    // written in order to write it to the batch response stream. So we need to ignore the Close()
+    // and provide an alternate method to release the stream after writing its content to the batch response.
+    internal class BatchStream : MemoryStream
+    {
+        /// <summary>
+        /// Release the batch stream and underlying resources
+        /// </summary>
+        internal void Release()
```
Comment (xuzhg) [inline src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs]: >Release() [](start = 22, length = 9)

Detach ?

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs [hunk 2]
```diff
@@ -241,6 +241,150 @@ public async Task CanHandleAbsoluteAndRelativeUrls()
...
+
+            string relativeToServiceRootUri = "DefaultBatchCustomer";
+            string relativeToHostUri = address.LocalPath.TrimEnd(new char[] { '/' }) + "/DefaultBatch/DefaultBatchCustomer";
+            string absoluteUri = this.BaseAddress + "/DefaultBatch/DefaultBatchCustomer";
+
+            // Act
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, requestUri);
+            request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json"));
+            HttpContent content = new StringContent(@"
+            {
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: >                                                                                                         [](start = 13, length = 104)

a lot of whitespaces are needless. Consider to remove all of them?

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/UnbufferedODataBatchHandlerTests.cs [hunk 3]
```diff
@@ -240,26 +240,223 @@ public async Task CanHandleAbsoluteAndRelativeUrls()
...
+            {
+                ODataModelBuilder builder = configuration.CreateConventionModelBuilder();
+                configuration.MapODataServiceRoute(
+                    "batch",
+                    "UnbufferedBatch",
+                    GetEdmModel(builder),
+                    new DefaultODataPathHandler(),
+                    ODataRoutingConventions.CreateDefault(),
+                    configuration.CreateUnbufferedODataBatchHandler());
+            }
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/UnbufferedODataBatchHandlerTests.cs]: revert these changes?

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs [hunk 4]
```diff
@@ -241,6 +241,150 @@ public async Task CanHandleAbsoluteAndRelativeUrls()
...
+            var requestUri = string.Format("{0}/DefaultBatch/$batch", this.BaseAddress);
+            Uri address = new Uri(this.BaseAddress, UriKind.Absolute);
+
+            string relativeToServiceRootUri = "DefaultBatchCustomer";
+            string relativeToHostUri = address.LocalPath.TrimEnd(new char[] { '/' }) + "/DefaultBatch/DefaultBatchCustomer";
+            string absoluteUri = this.BaseAddress + "/DefaultBatch/DefaultBatchCustomer";
+
+            // Act
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, requestUri);
+            request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json"));
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: >request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json")); [](start = 12, length = 86)

this line is mandatory?

File: src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+using System.IO;
+
+namespace Microsoft.AspNet.OData.Batch
+{
+    // Normally, the stream is closed when the ODataMessageWriter is disposed.
+    // For responses within a batch, we need to read from the stream after the response is
+    // written in order to write it to the batch response stream. So we need to ignore the Close()
+    // and provide an alternate method to release the stream after writing its content to the batch response.
+    internal class BatchStream : MemoryStream
```
Comment (xuzhg) [inline src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs]: >BatchStream [](start = 19, length = 11)

ODataBatchStream?

File: src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs (hunk 1 above)
Comment (xuzhg) [inline src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs]: >internal void Release() [](start = 8, length = 23)

shall we add a protect mechanism? 
how about to call twice "Release()"? 

Maybe test IsClosed() before call base.Close()?

Comment (mikepizzo) [reply src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs]: Well, it's not actually detaching anything; the memory stream is fully contained by the ODataBatchStream; it's just releasing the resources associated with (underlying) stream.

Maybe Free()?  Discard? Internal_Dispose?

---
In reply to: [377796792](https://github.com/OData/WebApi/pull/2048#discussion_r377796792) [](ancestors = 377796792)

Comment (mikepizzo) [reply src/Microsoft.AspNetCore.OData/Batch/ODataBatchStream.cs]: Went with InternalDispose(), since that's really what's happening.

---
In reply to: [377919530](https://github.com/OData/WebApi/pull/2048#discussion_r377919530) [](ancestors = 377919530,377796792)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs (hunk 4 above)
Comment (mikepizzo) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: In WebAPI, if you don't specify an accept header, the response is multipart/mixed.  We should maybe fix this...

---
In reply to: [377800365](https://github.com/OData/WebApi/pull/2048#discussion_r377800365) [](ancestors = 377800365)

Comment (mikepizzo) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: Updated WebAPI to set default to JSON if request is JSON and  no accept is specified.

---
In reply to: [377925870](https://github.com/OData/WebApi/pull/2048#discussion_r377925870) [](ancestors = 377925870,377800365)

File: src/Microsoft.AspNetCore.OData/Batch/ODataBatchResponseItem.cs [hunk 6]
```diff
@@ -17,6 +17,51 @@ namespace Microsoft.AspNet.OData.Batch
     /// </summary>
     public abstract class ODataBatchResponseItem
     {
+        /// <summary>
+        /// Writes a single OData batch response Synchronously.
+        /// </summary>
+        /// <param name="writer">The <see cref="ODataBatchWriter"/>.</param>
+        /// <param name="context">The message context.</param>
+        public static void WriteMessage(ODataBatchWriter writer, HttpContext context)
```
Comment (xuzhg) [inline src/Microsoft.AspNetCore.OData/Batch/ODataBatchResponseItem.cs]: >atic void WriteMessage(ODataBatchWriter writer, HttpContext con [](start = 17, length = 63)

can we call "WriteMessageAsync" in WriteMessage, or Call "WriteMessage()" in "WriteMessageAsync"?

Comment (mikepizzo) [reply src/Microsoft.AspNetCore.OData/Batch/ODataBatchResponseItem.cs]: I got rid of WriteMessage, and instead just created a WriteMessageAsync that took a flag to determine whether the writer was async.  When we have a major breaking change release of WebAPI OData I want to make all of this internal; there's no need for this functionality to be public, and it's confusing and broken in its current state.

---
In reply to: [378431693](https://github.com/OData/WebApi/pull/2048#discussion_r378431693) [](ancestors = 378431693)
